File: CameraTool.m
/*
Copyright (C) 2010-2012 Kristian Duske

This file is part of TrenchBroom.

TrenchBroom is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

TrenchBroom is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with TrenchBroom.  If not, see <http://www.gnu.org/licenses/>.
*/

#import "CameraTool.h"
#import "MapWindowController.h"
#import "Camera.h"
#import "PickingHit.h"
#import "PickingHitList.h"

@interface CameraTool (private)

- (BOOL)isCameraModifierPressed;
- (BOOL)isCameraOrbitModifierPressed;

@end

@implementation CameraTool (private)

- (BOOL)isCameraModifierPressed {
    return [NSEvent modifierFlags] == NSShiftKeyMask;
}

- (BOOL)isCameraOrbitModifierPressed {
    return [NSEvent modifierFlags] == (NSShiftKeyMask | NSCommandKeyMask);
}

@end

@implementation CameraTool

- (void)beginLeftDrag:(NSEvent *)event ray:(TRay *)ray hits:(PickingHitList *)hits {
    if (![self isCameraModifierPressed] && ![self isCameraOrbitModifierPressed])
        return;
    
    if ([self isCameraOrbitModifierPressed]) {
        PickingHit* hit = [hits firstHitOfType:HT_ANY ignoreOccluders:YES];
        if (hit != nil) {
            orbitCenter = *[hit hitPoint];
        } else {
            Camera* camera = [windowController camera];
            orbitCenter = [camera defaultPoint];
        }
        orbit = YES;
    }
}

- (void)endLeftDrag:(NSEvent *)event ray:(TRay *)ray hits:(PickingHitList *)hits {
    orbit = NO;
}

- (void)leftDrag:(NSEvent *)event ray:(TRay *)ray hits:(PickingHitList *)hits {
    Camera* camera = [windowController camera];
    if (orbit) {
        float h = -[event deltaX] / 90;
        float v = [event deltaY] / 90;
        [camera orbitCenter:&orbitCenter hAngle:h vAngle:v];
    } else {
        float yaw = -[event deltaX] / 90;
        float pitch = [event deltaY] / 90;
        [camera rotateYaw:yaw pitch:pitch];
    }
}

- (void)rightDrag:(NSEvent *)event ray:(TRay *)ray hits:(PickingHitList *)hits {
    if (![self isCameraModifierPressed])
        return;
    
    Camera* camera = [windowController camera];
    [camera moveForward:0 right:[event deltaX] up:-[event deltaY]];
}

- (void)handleBeginGesture:(NSEvent *)event ray:(TRay *)ray hits:(PickingHitList *)hits {
    gesture = YES;
}

- (void)handleEndGesture:(NSEvent *)event ray:(TRay *)ray hits:(PickingHitList *)hits {
    gesture = NO;
}

- (void)handleScrollWheel:(NSEvent *)event ray:(TRay *)ray hits:(PickingHitList *)hits {
    if (![self isCameraModifierPressed] && ![self isCameraOrbitModifierPressed])
        return;
    
    Camera* camera = [windowController camera];
    if ([self isCameraOrbitModifierPressed]) {
        if (gesture)
            [camera setZoom:[camera zoom] + [event deltaZ] / 10];
        else
            [camera setZoom:[camera zoom] + [event deltaX] / 10];
    } else {
        if (gesture)
            [camera moveForward:6 * [event deltaZ] right:-6 * [event deltaX] up:6 * [event deltaY]];
        else
            [camera moveForward:6 * [event deltaX] right:-6 * [event deltaY] up:6 * [event deltaZ]];
    }
}

- (void)handleMagnify:(NSEvent *)event ray:(TRay *)ray hits:(PickingHitList *)hits {
    if (![self isCameraModifierPressed])
        return;
    
    Camera* camera = [windowController camera];
    if ([self isCameraOrbitModifierPressed])
        [camera setZoom:[camera zoom] - [event magnification] / 2];
    else
        [camera moveForward:160 * [event magnification] right:0 up:0];
}

@end

File: Core/Controller/CameraTool.cpp
/*
 Copyright (C) 2010-2012 Kristian Duske
 
 This file is part of TrenchBroom.
 
 TrenchBroom is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.
 
 TrenchBroom is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License
 along with TrenchBroom.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "CameraTool.h"
#include <cstdio>
#include "Camera.h"
#include "Editor.h"
#include "Picker.h"

namespace TrenchBroom {
    namespace Controller {
        bool CameraTool::scrolled(ToolEvent& event) {
            if (!cameraModiferPressed(event) && !orbitModifierPressed(event)) return false;
            
            float forward = event.scrollX * m_moveSensitivity;
            float right = 0;
            float up = 0;
            m_editor.camera().moveBy(forward, right, up);
            return true;
        }
        
        bool CameraTool::beginLeftDrag(ToolEvent& event) {
            if (!cameraModiferPressed(event) && !orbitModifierPressed(event)) return false;
            
            if (orbitModifierPressed(event)) {
                Model::Hit* hit = event.hits->first(Model::HT_ENTITY | Model::HT_FACE, true);
                if (hit != NULL) m_orbitCenter = hit->hitPoint;
                else m_orbitCenter = m_editor.camera().defaultPoint();
                m_orbit = true;
            }
            return true;
        }
        
        void CameraTool::leftDrag(ToolEvent& event) {
            if (m_orbit) {
                float hAngle = -event.deltaX * m_lookSensitivity;
                float vAngle = event.deltaY * m_lookSensitivity * (m_invert ? 1 : -1);
                m_editor.camera().orbit(m_orbitCenter, hAngle, vAngle);
            } else {
                float yawAngle = -event.deltaX * m_lookSensitivity;
                float pitchAngle = event.deltaY * m_lookSensitivity * (m_invert ? 1 : -1);
                m_editor.camera().rotate(yawAngle, pitchAngle);
            }
        }
        
        void CameraTool::endLeftDrag(ToolEvent& event) {
            m_orbit = false;
        }
        
        bool CameraTool::beginRightDrag(ToolEvent& event) {
            return cameraModiferPressed(event) || orbitModifierPressed(event);
        }
        
        void CameraTool::rightDrag(ToolEvent& event) {
            float forward = 0;
            float right = event.deltaX * m_panSensitivity;
            float up = -event.deltaY * m_panSensitivity;
            m_editor.camera().moveBy(forward, right, up);
        }
        
        bool CameraTool::cameraModiferPressed(ToolEvent& event) {
            return event.modifierKeys == MK_SHIFT;
        }
        
        bool CameraTool::orbitModifierPressed(ToolEvent& event) {
            return event.modifierKeys == (MK_SHIFT | MK_CMD);
        }
    }
}